### ISSUE TITLE
Implement data storing

### ISSUE THREAD

**Issue by merkys** [EnjoyCpp/jsonapi-client-cpp#27]
So far the code only reads data from JSON API servers. We are going to need to send `Resource` data to servers. JSON API descriptions has [a well-written specification for that](https://jsonapi.org/format/#crud).

After #26 is implemented, a `Resource` will know its `Server`. A method `Resource::store()` has to be implemented (no parameters) that would do the following:
* Attempt to create a new remote resource if `id` does not exist (if successful, the `Resource` should store the remotely assigned `id`);
* Attempt to update a resource if `id` does not exist.

To test this, a sandbox JSON API server may be used: http://jsonapiplayground.reyesoft.com

**Comment by EnjoyCpp** [comment]
@merkys started to implement data storing. For some reason part URL of server is symbolized so now URL is written by hand

**Comment by merkys** [comment]
> @merkys started to implement data storing. For some reason part URL of server is symbolized so now URL is written by hand

Could you provide more details concerning the problem? Maybe I could help, but I need to understand it better.

**Edit:** Is this related to #32?

**Comment by EnjoyCpp** [comment]
@merkys The problem is that URL of server is something like this: �����Usonapiplayground.reyesoft.com/v2/authors/2.

**Comment by merkys** [comment]
OK, I will investigate. However, I cannot compile now due to #34.

**Comment by merkys** [comment]
> @merkys The problem is that URL of server is something like this: �����Usonapiplayground.reyesoft.com/v2/authors/2.

This is due to #32. Currently the code tries to convert `Server` object to string, and this must not be done.

**Comment by EnjoyCpp** [comment]
@merkys is my Resource::store() somewhat correctly? Because right now i am a little bit lost what i have to do next.

**Comment by merkys** [comment]
> @merkys is my Resource::store() somewhat correctly? Because right now i am a little bit lost what i have to do next.

It looks mostly OK. A couple of issues spotted in the code:

https://github.com/EnjoyCpp/jsonapi-client-cpp/blob/bdf6b284bb8d78d4e671891e2269a520c2ef37ca/src/Resource.cpp#L81

Server URL should be taken via `get_url()` method of `Server` object. Symbolization will most likely fail.

https://github.com/EnjoyCpp/jsonapi-client-cpp/blob/bdf6b284bb8d78d4e671891e2269a520c2ef37ca/src/Resource.cpp#L84

Not needed for now. Later on a field `authorization` could be added for `Server` (with its own setter/getter pair, default value `NULL`), which could be consulted to add the header.

Moreover, I guess `CURLOPT_POSTFIELDSIZE` should be set for successful POST (see [here](https://curl.haxx.se/libcurl/c/CURLOPT_POSTFIELDS.html) for example).

It would be great if you could add a test under `tests/` for store functionality, using http://jsonapiplayground.reyesoft.com/v2 to write data. This way it would be much easier for us to check whether storing functionality really works as expected.

**Comment by EnjoyCpp** [comment]
@merkys how should I test my store() functionality? Like checking response code or is curl is good ?

**Comment by merkys** [comment]
> @merkys how should I test my store() functionality? Like checking response code or is curl is good ?

HTTP response codes should be checked to see whether `store()` succeeded. Please consult https://jsonapi.org/format/#crud to see what to expect (mostly HTTP `2xx` codes, others will inform about errors).

**Comment by EnjoyCpp** [comment]
@merkys Http response code i get from "http://jsonapiplayground.reyesoft.com/v2/authors/2" is 405 (405 Method Not Allowed) . Also changed method Resource::store() from void.

**Comment by merkys** [comment]
> @merkys Http response code i get from "http://jsonapiplayground.reyesoft.com/v2/authors/2" is 405 (405 Method Not Allowed) . Also changed method Resource::store() from void.

This is because `/authors/2` already exists. Now I am thinking that `store()` should be split to two separate methods:
* `create()` should POST data to `/<type>`;
* `update()` should PATCH data to `/<type>/<id>`.

Previously I thought we could get around with having a generic `store()`, however, we would have to somehow detect which `Resource` is downloaded, and which is to be created anew. So it's better to have two separate methods.

**Comment by EnjoyCpp** [comment]
@merkys i separated store to POST and PATCH methods and added tests to both of them.

**Comment by merkys** [comment]
Good. However, both tests for create and update fail on my machine. Somewhy server URL is hard-coded in both `create()` and `update()`. Please replace them with URLs derived from `Server` object, just as indicated earlier.

Method `create()` may be failing due to the fact that `id` field is specified in the POSTed JSON. In the test for `create()`, `id` should be unset prior creating the resource, otherwise it will conflict with already existing entry. To unset `id`, a method `unset_id()` should be added for `Resource`. This method should clear the existing value of `id` in `Resource`'s JSON.

**Comment by EnjoyCpp** [comment]
@merkys i have added unset_id() method. URLs derived from Server object returns wrong value so maybe i have to convert them?

**Comment by merkys** [comment]
> @merkys i have added unset_id() method. URLs derived from Server object returns wrong value so maybe i have to convert them?

The problem with server URL is `Server::get_one()`:
https://github.com/EnjoyCpp/jsonapi-client-cpp/blob/416e31fc6225402cc93de71c0c0efc4c5713eac0/src/Server.cpp#L91-L92

A new `Server` is created, while the old server must be used, just as is done in `Server::get_all()`. Using the same server should not cause any problems.

Methods `create()` and `update()` must use the URLs as written in my previous comment (https://github.com/EnjoyCpp/jsonapi-client-cpp/issues/27#issuecomment-629937066), as now they seem to send requests for base server URL.

**Comment by EnjoyCpp** [comment]
@merkys is Server::get_one() method now is good? Because problem seems to be the same.

**Comment by merkys** [comment]
> @merkys is Server::get_one() method now is good? Because problem seems to be the same.

I see no new commits since 3772d4b162aac64d2f7efa978a5d987ba736f522. Maybe you have forgotten to push yours?

**Comment by EnjoyCpp** [comment]
@merkys Yes i forgot. Sorry about that.

**Comment by merkys** [comment]
> @merkys is Server::get_one() method now is good? Because problem seems to be the same.

Yes, `Server::get_one()` is fine now. `Server`s are still created in `get_one()` and `get_all()`, although they are unused - these lines of code should be removed.

Methods `create()` and `update()`, however, still use incorrect URLs.

**Comment by EnjoyCpp** [comment]
@merkys is this `server->get_url()` is a correct way to get URL of the server? Because program always tries to rebuilt URL. 
https://github.com/EnjoyCpp/jsonapi-client-cpp/blob/a33fa90889dd99907c9243b3ffca2a379c98c74d/src/Resource.cpp#L83

**Comment by merkys** [comment]
> @merkys is this `server->get_url()` is a correct way to get URL of the server? Because program always tries to rebuilt URL.

Fixed in #44, please examine the changes and check why tests report failures.

**Comment by EnjoyCpp** [comment]
@merkys now tests are good i think. Also i spotted  one mistake when i try to set value of attribute using string type it shows wrong value.
For example trying to set value for attribute name ( `set_attribute("name","Pablo Picasso")`);
In JSON shows `"name" : true`
And in the POSTED data to the website  `name: "1".`
So think there is a mistake in this setter that i will try to solve:
https://github.com/EnjoyCpp/jsonapi-client-cpp/blob/feac235d59dc55f374c801fc25b2b3702b87c602/src/Resource.cpp#L27-L28

**Comment by merkys** [comment]
> For example trying to set value for attribute name ( `set_attribute("name","Pablo Picasso")`);
> In JSON shows `"name" : true`
> And in the POSTED data to the website `name: "1".`

This seems to be because `set_attribute( const std::string name, bool value )` gets called for `set_attribute("name","Pablo Picasso")`, not sure why. I tried commenting out the `bool` setter and it worked fine. Maybe adding a `const char *` setter would work?

**Comment by merkys** [comment]
Somewhy tests still exit with non-zero exit statuses, although their results show successful create/update.

**Comment by EnjoyCpp** [comment]
@merkys now test should work correctly

**Comment by merkys** [comment]
Great! Thus I close this issue, as storing works now. There are minor issues, but I will report them as separate GitHub issues.
